libraries/AP_Scripting/examples/Serial_Dump.lua

```diff
@@ -0,0 +1,44 @@
+-- this script reads data from a serial port and dumps it to a file
+local file_name = 'serial dump.txt'
+local baud_rate = 9600
+
+-- find the serial first (0) scripting serial port instance
+local port = serial:find_serial(0) -- SERIALx_PROTOCOL 28
+
+if not port then
+    error("Could not find Scripting Serial Port")
+    return
+end
+
+-- make a file
+file = io.open(file_name, "w")
+if not file then
+  error("Could not create file " .. file_name)
+end
+file:close()
+
+-- begin the serial port
+port:begin(baud_rate)
+port:set_flow_control(0)
+
+function update() -- this is the loop which periodically runs
+
+  local n_bytes = port:available()
+  local string = ''
+  while n_bytes > 0 do
+    -- concatenate all bytes into string
+    string = string .. port:read()
```

Comment:
Not critical here, as this works for a simple/low data rate port, but on a fast link this is going to be both inefficent, and can be very prone to memory problems, as you are building a new string each call, so you are paying the overhead of copying the string for each byte, and if the string is sufficiently long enough we will have to store 2n - 1 worth of strings at the same time, which we may not be able to allocate due to fragmentation.

A better approach here is probably to have an outer and inner loop, where the inner loop won't copy more then say 512 bytes (or similar) before writing to the disk, which would limit your worse case memory consumption. You could also potentially use a fixed size table then pack to a string at the final step.

If you know your system won't be overloaded, that the data is coming in low enough rate then what's here is probably fine, but it won't be robust on memory constrained systems/high baud rates, so it's probably worth demoing a way to limit the memory we burn here.

Reply:
fixed

Reply:
I'd just add an outer loop to your loop there. So you have a loop that looks like this:

```
remaining = available_bytes()
while remaining > 0
  read_bytes = 0
  bytes_to_read = MIN(remaining, 512)
  while read_bytes < bytes_to_read
    read a byte to buffer
  write buffer to file
  remaining = remaining - read_bytes
```

Then you shouldn't have to change the overall rate

Reply:
You right, that is better, I have done that and re-tested